### PR TITLE
Add TermStoreInstaller interface

### PR DESCRIPTION
**Pull request by JeroenDeDauw**

### CODE REVIEW

**Comment by JeroenDeDauw** [comment]
@AlaaSarhan I noticed not having this interface in Wikibase Repo. Without it there will be some implementation dependent code outside of the wiring part of the application. So outcome of this little experiment for me is: keep dedicated installer interface in components like this

**Comment by JeroenDeDauw** [comment]
Please bikeshed about the naming and structure here, while it is still trivial to change :)

File: src/TermStoreInstaller.php [hunk 1]
```diff
@@ -0,0 +1,17 @@
+<?php
+
+namespace Wikibase\TermStore;
+
+interface TermStoreInstaller {
+
+	/**
+	 * @throws TermStoreException
+	 */
+	public function installOrUpdate();
```
Comment (AlaaSarhan) [inline src/TermStoreInstaller.php:10]: on naming this `installOrUpdate`.. this comes as a bit weird thing to assume in an abstraction. like why assume there are two different behaviors `install` or `update` here. If they can be distinguished on abstract level here, they need to be separate probably. If not, I would just call it `install` and let implementations decide what it means for each of them

File: src/TermStoreInstaller.php [hunk 2]
```diff
@@ -0,0 +1,17 @@
+<?php
+
+namespace Wikibase\TermStore;
+
+interface TermStoreInstaller {
+
+	/**
+	 * @throws TermStoreException
```
Comment (AlaaSarhan) [inline src/TermStoreInstaller.php:8]: I would remove interface exceptions entirely .. will do that in a separate PR to discuss there

**Review by AlaaSarhan — COMMENTED** [review]
added two bikeshed opinions

File: src/TermStoreInstaller.php (hunk 1 above)
Comment (JeroenDeDauw) [reply src/TermStoreInstaller.php:10]: I arrived at `installOrUpdate` from the user side, not from the implementer side. update.php is "update to latest version or install if it is not installed yet". So to me it seems like either we need a method like this, or expose more stuff to the caller can call `install` or `update`. Do you see another option?

Comment (AlaaSarhan) [reply src/TermStoreInstaller.php:10]: "update to latest version or install if it is not installed yet" sounds a bit concrete for this abstraction level. In databases, or things that persist long-term involving schemas, it might makes sense but that's not the case in other kinds of systems (file systems, key-value storages, in-memory caching mechanism ... etc).

so I would call this install, and if it means in some implementation that it creates or updates a table then let it be that way there.

Comment (JeroenDeDauw) [reply src/TermStoreInstaller.php:10]: You are right there is not always an update. At the moment there is none for the Doctrine implementation, and indeed for some implementations you will never have one. The same can be said about install for stuff like in memory solutions.

The standard approach to interface design is to look at what the consumers need. Originally I figured there would be no sensible users when we where talking about if we need an interface like this at all. Turns out I was wrong and that we do have one. This consumer needs a way to tell the store to do any install and or update to get to its latest version. I see no problems from either a design or practical perspective if for some implementations this is a no-op.

I do OTOH see a problem with having an `install` method that also does updating. This is surprising. And from the interface you cannot tell this behavior is supported.

Comment (JeroenDeDauw) [reply src/TermStoreInstaller.php:10]: > The standard approach to interface design is to look at what the consumers need. 

I'd like to stress this point. A lot of people do not have this in mind. Interfaces are about their consumers, not their implementers.

Comment (AlaaSarhan) [reply src/TermStoreInstaller.php:10]: > I do OTOH see a problem with having an install method that also does updating. This is surprising. And from the interface you cannot tell this behavior is supported.

but here you are still assuming that install and update are two distinct behaviors from a consumer that cares about the abstraction. If that is the case, then they should be separate.

>>    The standard approach to interface design is to look at what the consumers need.

> I'd like to stress this point. A lot of people do not have this in mind. Interfaces are about their consumers, not their implementers.

1. For that reason exactly I do think `installOrUpdate` is not thinking about all consumers. A consumer who wants to depend on this abstraction, does not want to worry about implementation details thus they do not even want to know what install means for whatever implementation is configured. So `install` is sufficient.

2. If you think about implementations on the other hand, which you also should do the way you do for consumers (and balance the too in a decent manner), giving me something that will be a no-op operation in my case is not desirable (that is not what abstraction is about anyway). Of course unless that's really necessary and it is a matter of edge case. I do not see here necessity nor edge-cases that will face this, rather a big chunk of implementers.

From 1 and 2 I conclude that the decent decision here is not to assume there are two different behaviors yet.

If you really disagree and think there's a good reason for distinction, then there should be a new concept in this abstraction, either a separate class `TermStoreUpdater` or a separate method `TermStoreInstaller::update()`.

To my mind, and I think it is a good interface design principle (also do not know if it has been stated by some other genius people, so I'll go ahead and call it No Half-Concepts Principle (NHCP) :p sounds like some hokey league name now ):
> one should not do half-way decisions in interfaces.. either a concept exist or not

----

Though, I do not want to strongly disagree, I am approving this for now and let it be an experiment of that principle (I'm happy to be proven wrong by receiving no complaints on the name in 20 years from now, well probably not worth complaining about in the first place, thus the approval ;) )

**Review by AlaaSarhan — APPROVED** [review]

**Review by AlaaSarhan — APPROVED** [review]
@JeroenDeDauw feel free to merge at will

**Comment by AlaaSarhan** [comment]
One a second thought, while working on MW DBAL impl .. maybe you do not need to abstract installation related stuff at all after all .. installation anyway is usually handled outside of application logic (consumers of this interface and it's implementations) but somewhere else in the stack/framework. There configured implementations' ways of installations can be accessed directly and that should also be fine from design perspective in general.

File: src/TermStoreInstaller.php (hunk 1 above)
Comment (JeroenDeDauw) [reply src/TermStoreInstaller.php:10]: For more clarity we should probably talk about this in person. Feel free to merge already is you deem that to be useful.

**Comment by JeroenDeDauw** [comment]
cat-chasing-tail

You are back to where I was last week. Please see my first comment in this PR.

**Comment by AlaaSarhan** [comment]
yeap would like to discusss this in person.
I think it isn't necessary yet so won't merge for now.

**Comment by JeroenDeDauw** [comment]
We will put similar interface in WB repo
